### DISCUSSION TITLE
Installing Guide Firefly-iii on IIS with MariaDB  (Windows)

### DISCUSSION THREAD

**Discussion by TheFork001** [firefly-iii/firefly-iii#8400] [General]
Alright for anyone trying to Install this on IIS, this is how far I have gotten. I hope this helps everyone as it was a bit of error and trial on my part. 

Step one: install PHP NTS or TS on iis server follow other guides on this part.

Step Two: PHP BCMath Arbitrary Precision Mathematics, PHP Internationalization extension, PHP Curl, PHP Zip, PHP Sodium, PHP GD, PHP XML, PHP MBString are all in the php.ini file make sure to un-commment them. Their name varies in windows so do a bit of research, they are all there I promise. **Important: Under the [curl] tag you will see (;curl.cainfo =) uncomment it go to https://curl.haxx.se/ca/cacert.pem and download it, place it in the php folder and make sure you type the path directly to it. Example: curl.cainfo = C:\PHP\cacert.pem ** 

Step three: Remember to set up the correct permissions on the service account in IIS for that root directory and make sure iis is pointing to the "public" folder of course and the url rewrite rules are going to be imported automatically assuming you have the URL rewrite module installed. If you don't go search for it and install it.

Step Four: Install composer for php, there are many guides and easy to find if it's not already installed. I then used this code: `curl -sS https://getcomposer.org/installer | sudo php -- --install-dir=/usr/local/bin --filename=composer` it can be found here: https://docs.firefly-iii.org/how-to/firefly-iii/installation/self-managed/ and installed firefly-iii

Step Five: Back out to the wwroot/firefli-iii right click that folder, click properties, and go to permissions find the account or group your service account is under and add it with all of the rights or if it's already there, just check a box and uncheck it, click apply, all permissions will be fixed, then click ok. 

Step Six: Configure the database file under mysql option in the .env file, set DEFAULT_LANGUAGE=English and **VERY Important: when changing the language to "English" you have to go find the en_US folder under "firefly-iii\resources\lang" and change the folder name to English.** When setting up email, If you use gmail, the app password has spaces so add it like this "sdf ghj gds fgh".

Step Seven: Assuming you set these settings right then when you run `php artisan firefly-iii:upgrade-database` within the firefly-iii directory the database will start getting populated. I then run This guy: `php artisan firefly-iii:correct-database` and then `php artisan firefly-iii:report-integrity` and go log in. All features should work accordingly!

PS: I changed my LOG_CHANNEL=errorlog in .env to "errorlog". Personal preference.

As far as your headers go, thats up to you but if the HTTP response headers are set in iis correctly you should also not have any issues.

**If you want SSL settings on your config, adding it to the .env file in iis did not work for me, I ended up adding it in the .env a file and the database config file located under firefly-iii\config I replaced the values under the // Get SSL parameters in the database config file. Then the db worked just fine with SSL enabled**

**Note:** Don't touch the language in your profile, it's set already. If you change it to English US (global) it will break, leave it as is. 

I hope my elementary guide helps anyone trying to install this in IIS windows, my guide worked for me. I did not see one and through error and trial, I got everything going in iis. I hope this will help you, if it did heart it :)

**Comment by JC5** [comment]
Nice, thank you!

**Reply by TheFork001** [reply]
You're welcome. As soon as I finish everything to the best of my abilities, I will retype this and make it very detailed for windows and iis users to use.

**Reply by TheFork001** [reply]
Done! @JC5 everything is working as it should now!

**Reply by TheFork001** [reply]
I tried adding my ssl settings to the .env file but for some reason, it will not take the cert paths when attempting to connect to the environment. I had to add them to the database ssl config file manually and it worked. If I dont do that I get the error if not I get the error "SQLSTATE[HY000] [2002] Cannot connect to MySQL using SSL (Connection: mysql, SQL: select count(*) as aggregate from `users`)"

**Reply by JC5** [reply]
The cert paths? Could you elaborate?

**Reply by TheFork001** [reply]
Yes, so in the section below of the .env file it wants the path to the ca, cert,and key. I added it there but in the database file it says it will get the path from the .env file. It does not.

MYSQL_USE_SSL=true
MYSQL_SSL_VERIFY_SERVER_CERT=true
MYSQL_SSL_CAPATH=PATH to my ssl
MYSQL_SSL_CA=PATH to my ssl
MYSQL_SSL_CERT=PATH to my ssl
MYSQL_SSL_KEY=PATH to my ssl
MYSQL_SSL_CIPHER=

Only when I set the path in the database config file it works unless I'm doing something wrong.

// Get SSL parameters from .env file.
$mysql_ssl_ca_dir  = envNonEmpty('PATH to my ssl', null);
$mysql_ssl_ca_file = envNonEmpty('PATH to my ssl', null);
$mysql_ssl_cert    = envNonEmpty('PATH to my ssl', null);
$mysql_ssl_key     = envNonEmpty('PATH to my ssl', null);
$mysql_ssl_ciphers = envNonEmpty('MYSQL_SSL_CIPHER', null);
$mysql_ssl_verify  = envNonEmpty('MYSQL_SSL_VERIFY_SERVER_CERT', null);

**Reply by JC5** [reply]
It does get it from there, the code is pretty unambiguous:

```php
// this works if the env var is set
$mysql_ssl_ca_dir  = envNonEmpty('MYSQL_SSL_CAPATH', null);
```

If you change this to what you proposed, with the literal path in the place of the environment variable name, it should not work at all.

```php
// this should not work at all. 
 $mysql_ssl_ca_dir = envNonEmpty('c:/blabla', null);
```

But perhaps you need to experiment:

```
MYSQL_SSL_CAPATH="c:/blabla/"
MYSQL_SSL_CAPATH=c:/blabla/
MYSQL_SSL_CAPATH="c:\/blabla\/"
```

**Reply by TheFork001** [reply]
Trying these variations now, I tried all of those, and nothing. only when I place it in the database.php file it works. Ill keep trying.
